chia/wallet/puzzles/p2_singleton.clvm

```diff
@@ -0,0 +1,61 @@
+(mod (SINGLETON_MOD_HASH SINGLETON_MOD_HASH_HASH SINGLETON_ID singleton_innerpuzhash my_id)
+
+  (include condition_codes.clvm)
+
+  (defun is-in-list (atom items)
+    ;; returns 1 iff `atom` is in the list of `items`
+    (if items
+      (if (= atom (f items))
+        1
+        (is-in-list atom (r items))
+      )
+      0
+    )
+  )
+
+  ;; utility function used by `curry_args`
+  (defun fix_curry_args (items core)
+    (if items
+        (qq (c (q . (unquote (f items))) (unquote (fix_curry_args (r items) core))))
+        core
+    )
+  )
+
+  ; (curry_args sum (list 50 60)) => returns a function that is like (sum 50 60 ...)
+  (defun curry_args (func list_of_args) (qq (a (q . (unquote func)) (unquote (fix_curry_args list_of_args (q . 1))))))
+
+  ;; (curry sum 50 60) => returns a function that is like (sum 50 60 ...)
+  (defun curry (func . args) (curry_args func args))
+
+  ;; hash a tree with escape values representing already-hashed subtrees
+  ;; This optimization can be useful if you know the puzzle hash of a sub-expression.
+  ;; You probably actually want to use `curry_and_hash` though.
+  (defun sha256tree_esc_list
+         (TREE LITERALS)
+         (if (l TREE)
+             (sha256 2 (sha256tree_esc_list (f TREE) LITERALS) (sha256tree_esc_list (r TREE) LITERALS))
+             (if (is-in-list TREE LITERALS)
+                 TREE
+                 (sha256 1 TREE)
+             )
+         )
+  )
+
+  ;; hash a tree with escape values representing already-hashed subtrees
+  ;; This optimization can be useful if you know the tree hash of a sub-expression.
+  (defun sha256tree_esc
+     (TREE . LITERAL)
+     (sha256tree_esc_list TREE LITERAL)
+  )
+
+  ;; return the puzzle hash for a cc with the given `genesis-coin-checker-hash` & `inner-puzzle`
+  (defun-inline create_fullpuzhash (mod_hash mod_hash_hash genesis_id inner_puzzle_hash)
+    (sha256tree_esc (curry mod_hash mod_hash_hash genesis_id inner_puzzle_hash)
+                    mod_hash
+                    mod_hash_hash
+                    inner_puzzle_hash)
+  )
+
+  ; main
+  (list (list ASSERT_PUZZLE_ANNOUNCEMENT (sha256 (create_fullpuzhash SINGLETON_MOD_HASH SINGLETON_MOD_HASH_HASH SINGLETON_ID singleton_innerpuzhash) my_id)) (list CREATE_COIN_ANNOUNCEMENT 0x00) (list ASSERT_MY_ID my_id))
+)
```

Comment:
Why `CREATE_COIN_ANNOUNCEMENT 0x00`? What's the intent of this announcement?

Reply:
It's a generic announcement. We don't actually care about the message, we just want to be able to assert that this coin has been spent